Added get_param staticmethod for augmentations

shijianjian opened the pull request.

shijianjian commented:
https://github.com/kornia/kornia/issues/312#issuecomment-570113696

edgarriba commented on kornia/augmentation/augmentations.py:
```diff
@@ -1,21 +1,34 @@
-from kornia.color.adjust import AdjustBrightness, AdjustContrast, AdjustSaturation, AdjustHue
-from kornia.color.gray import rgb_to_grayscale
-from kornia.geometry import pi
+from kornia.augmentation.functional import *
```
I tend  to import what I really need, otherwise could be difficult to track used functions below in the code. Import as F or just the ones you need. Besides, if we want to behave with similar torch philosophy, we might abstract to another augmentation.utils/params module all the non exposed functions

edgarriba commented on kornia/augmentation/augmentations.py:
```diff
@@ -1,21 +1,34 @@
 
-class RandomFlip(nn.Module):
+class AugmentationBase(nn.Module):
+    def __init__(self, return_transform: bool = False):
```
mypy rules say that class constructors should return None

edgarriba commented on kornia/augmentation/augmentations.py:
```diff
@@ -24,7 +37,18 @@ def __repr__(self) -> str:
-    def forward_flip(self, input: UnionType, flip_func: Callable) -> UnionType:  # type: ignore
+    @staticmethod
+    def get_params(batch_size: int, p: float = 0.5):
+        return get_random_p_params(batch_size, p)
```
I would declare this function as `_get_random_p_params` since it won't be part of the exposed api

edgarriba commented on kornia/augmentation/augmentations.py:
```diff
@@ -85,12 +109,10 @@ class RandomHorizontalFlip(RandomFlip):
 
-    def forward(self, input: UnionType) -> UnionType:  # type: ignore
-        return self.forward_flip(input, random_hflip)
+    def forward(self, input: UnionType, params: dict = None) -> UnionType:  # type: ignore
```
mypy recommends to use Optional for None parameters

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+    return flipped
+
+
+def apply_vflip(input: torch.Tensor, params: dict, return_transform: bool = False) -> UnionType:
```
mark with underscore since it won't be part of the main api

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+    return flipped
+
+
+def apply_color_jitter(input: torch.Tensor, params: dict, return_transform: bool = False) -> UnionType:
```
mark with underscore since it won't be part of the main api

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+    return jittered
+
+
+def get_color_jitter_params(batch_size: int, brightness: FloatUnionType = 0.,
```
mark with underscore since it won't be part of the main api

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+    }
+
+
+def apply_grayscale(input: torch.Tensor, params: dict, return_transform: bool = False) -> UnionType:
```
mark with underscore since it won't be part of the main api

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+    return grayscale
+
+
+def get_random_p_params(batch_size: int, p: float = .5) -> dict:
```
mark with underscore since it won't be part of the main api

edgarriba commented on kornia/augmentation/augmentations.py:
```diff
@@ -125,15 +147,13 @@ class RandomVerticalFlip(RandomFlip):
 
-    def forward(self, input: UnionType) -> UnionType:  # type: ignore
-        return self.forward_flip(input, random_vflip)
+    def forward(self, input: UnionType, params: dict = None) -> UnionType:  # type: ignore
```
using mypy we can be more specific about the expected type like: `Optional[Dict[str, torch.Tensor]] = None`

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+
+    device: torch.device = input.device
+    dtype: torch.dtype = input.dtype
+    # print(input.detach().shape)
```
remove commented code

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+    else:
+        batch_size = input.shape[0] if len(input.shape) == 4 else 1
+    params = get_random_p_params(batch_size, p=p)
+    return apply_hflip(input, params, return_transform)
```
one question it comes to my mind, I understand the abstraction of the api, however, maybe we could follow the pattern below:
- core level -> `kornia.hflip` (exposed)
- augmentation level -> `kornia.random_hflip` (exposed)
  - this needs rand random parameters
    - call `_get_random_p_params`  and `_apply_hflip` which internally calls `kornia.hflip`

same for the other functions. Basically, we should have like a top level api, and some specifics for augmentation that uses random parameters generators. Maybe /cc  @fmassa could help with this

shijianjian replied:
In my original thoughts, I prefer to mark the ```random_*``` APIs as deprecated. I just rewrote them for passing the tests.

I would only like to preserve:
- The functional API, the F in your example, to do the augmentations in command.
- The torch module-like augmentations, for doing the augmentations on the fly.

edgarriba replied:
I don't think we are going to deprecate methods at this moment. In order to proceed with this refactoring we might need more consensus about the API. I understand the need to reproduce results and I think it's necessary but first we should have an initial version of the module as close as possible to torchvision interface. We are going to do a release this week including the methods left in the roadmap. As a second iteration, we can sketch out about how to generalise code. We could open an issue for that in order to have a clear Idea about the whole thing before starting to code and make major changes in the API.

ducha-aiki replied:
Well, we _may_ think about deprecating random_* stuff later. But I like that I can do things both functional or Module way now. And I like the way it is done in current PR: split functionality into "generate random" and "apply random". Both might be called either from Module or from functional.

edgarriba replied:
in that case, I would add extra functions that accept the generated parameters. Similar to what has been done here https://github.com/kornia/kornia/pull/399. And probably, in terms of reproducibility I would provide such methods as functionals.

shijianjian replied:
Right, it makes sense to make everything work in the first place. In that case, is it better to split the current functional.py into two separated files:
- functional.py, to include all the functional operations.
- param_gen.py, to include all the random param generators.

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,342 @@
+        batch_size = input[0].shape[0] if len(input[0].shape) == 4 else 1
+    else:
+        batch_size = input.shape[0] if len(input.shape) == 4 else 1
+    params = get_random_p_params(batch_size, p=p)
```
type for params to understand for code reader what's the returned data type

shijianjian commented:
I have made some further changes and refactoring:
1. Updated more strict typings according to @edgarriba 's review.
2. Split out param_gen.py out of functional.py.
3. Standardized the naming of param_gen.py to ```_random_*_gen``` like ```_random_prob_gen``` and ```_random_color_jitter_gen```.

edgarriba commented on kornia/augmentation/augmentations.py:
```diff
@@ -160,26 +182,34 @@ def __repr__(self) -> str:
+            if isinstance(input, tuple):
+                batch_size = input[0].shape[0] if len(input[0].shape) == 4 else 1
+            else:
+                batch_size = input.shape[0] if len(input.shape) == 4 else 1
```
suggest to add a function to infer the batch size since is used in every transform

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+    """Generate params and apply operation on input tensor.
+
+    See :class:`~kornia.augmentation.param_gen._random_prob_gen` for details.
+    See :class:`~kornia.augmentation.functional.apply_hflip` for details.
```
this are not classes. Check for another appropriate sphinx tags

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+    """Generate params and apply operation on input tensor.
+
+    See :class:`~kornia.augmentation.param_gen._random_prob_gen` for details.
+    See :class:`~kornia.augmentation.functional.apply_vflip` for details.
```
sphinx tag

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+    """Generate params and apply operation on input tensor.
+
+    See :class:`~kornia.augmentation.param_gen._random_color_jitter_gen` for details.
+    See :class:`~kornia.augmentation.functional.apply_color_jitter` for details.
```
sphinx

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+        input = input.unsqueeze(0)
+
+    if len(input.shape) == 3:
+        input = input.unsqueeze(0)
```
```python
    if len(input.shape) in (2, 3,):
        input = input.unsqueeze(0)
```

edgarriba replied:
however not quite sure about adding dimensions since later we do: ` input.view((-1, (*input.shape[-3:])))`

shijianjian replied:
Here, I think it is better to remove the code below to be more clear that we will only accept (H, W), (C, H, W) and (*, C, H, W):
```python
    input = input.unsqueeze(0)
    input = input.view((-1, (*input.shape[-3:])))
```

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+
+    Args:
+        params (dict): A dict that must have {'batch_prob': torch.Tensor}. Can be generated from
+        kornia.augmentation.param_gen._random_prob_gen
```
end '.'

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+    Returns:
+        torch.Tensor: The horizontally flipped input
+        torch.Tensor: The applied transformation matrix :math: `(*, 3, 3)` if return_transform flag
+                      is set to ``True``
```
end '.'

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+
+    flipped: torch.Tensor = input.clone()
+
+    to_flip = params['batch_prob'].to(device)
```
`.to(device)` could happen inside the generator

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+    return flipped
+
+
+def apply_vflip(input: torch.Tensor, params: Dict[str, torch.Tensor], return_transform: bool = False) -> UnionType:
```
similar to end point string as above and squeeze

edgarriba commented on kornia/augmentation/param_gen.py:
```diff
@@ -0,0 +1,118 @@
+        saturation (float or tuple): Default value is 0
+        hue (float or tuple): Default value is 0
+
+    See :class:`~kornia.augmentation.ColorJitter` for details.
```
not sure this should be here

edgarriba commented on kornia/augmentation/functional.py:
```diff
@@ -0,0 +1,286 @@
+
+def apply_grayscale(input: torch.Tensor, params: Dict[str, torch.Tensor], return_transform: bool = False) -> UnionType:
+    """Apply Gray Scale on a tensor image or a batch of tensor images with given random parameters.
+    Input should be a tensor of shape (3, H, W) or a batch of tensors :math:`(*, 3, H, W)`.
```
line break

edgarriba commented on kornia/augmentation/param_gen.py:
```diff
@@ -0,0 +1,118 @@
+
+def _random_color_jitter_gen(batch_size: int, brightness: FloatUnionType = 0.,
+                             contrast: FloatUnionType = 0., saturation: FloatUnionType = 0.,
+                             hue: FloatUnionType = 0.) -> Dict[str, torch.Tensor]:
```
device and dtype could be passed as input

edgarriba commented on kornia/augmentation/param_gen.py:
```diff
@@ -0,0 +1,118 @@
+    if not isinstance(p, float):
+        raise TypeError(f"The probability should be a float number. Got {type(p)}")
+
+    probs: torch.Tensor = torch.empty(batch_size, device='cpu').uniform_(0, 1)
```
pass device and maybe dtype

shijianjian commented on kornia/augmentation/augmentations.py:
```diff
@@ -1,21 +1,35 @@
+        repr = f"(return_transform={self.return_transform})"
+        return self.__class__.__name__ + repr
+
+    def forward(self, input: UnionType) -> UnionType:  # type: ignore
```
This is definitely a good suggestion to have a better abstraction of those functions.

For param generator part, I am thinking of something like:
```python
class ColorJitterParamDictGenerator():
      def random(a, b, c, d) -> dict:
             # 1. Generate random params
             # 2. Put params together
            return compose(_a, _b, _c, _d)
      def compose(a, b, c, d) -> dict:
            # This will be used for static params.
            return param_dict
      def validate(param_dict) -> bool:
            return valid_or_not
```
Also, I think the param generator should not do any ```.to(device).to(dtype)``` operations. The generator generates valid parameters, which is not device sensitive. Plus, in multi-gpus settings, one set of parameters may be used in different GPUs, which means its totally depending on the device and dtype of the batch images.

shijianjian commented:
This should be alright to merge. Problem with 'build-doc' in Python 3.6. Looks like it cannot find the kornia module. Cannot figure out what is the best solution.

edgarriba commented:
@shijianjian you can try to replicate the environment with python 2.6 using this argument 
https://github.com/kornia/kornia/blob/master/setup_dev_env.sh#L38

edgarriba commented on kornia/augmentation/__init__.py:
```diff
@@ -1,15 +1,20 @@
 
 __all__ = [
     "random_hflip",
+    # "apply_hflip",
```
remove commented code

edgarriba commented on kornia/augmentation/augmentations.py:
```diff
@@ -1,57 +1,52 @@
+        self._apply_fcn: Callable = apply_fcn
 
-    def forward_flip(self, input: UnionType, flip_func: Callable) -> UnionType:  # type: ignore
+    def infer_batch_size(self, input):
```
typing

edgarriba commented on requirements.txt:
```diff
@@ -23,7 +23,7 @@ mypy_extensions==0.4.1
 sphinx==2.1.2
 sphinx-gallery==0.4.0
-sphinx_rtd_theme=0.4.3
+sphinx_rtd_theme==0.4.3
```
could be this why documentation doesn't work with python 3.6 ?

shijianjian replied:
It wasn't that. It should be just a typo or it could not pass the static check by pip.

The doc can be built on my local machine by adding the module path with
```python
sys.path.append(current_directory)
```.

But no luck on the 3.6 CI. Can you test on your machine if possible?

Thanks,
Jian

edgarriba replied:
Will do during the day

shijianjian commented:
Hi @edgarriba @dvd42 

Thanks for your comments and fixes. Please check out the changes here.

edgarriba commented:
merged via https://github.com/kornia/kornia/pull/407

edgarriba commented:
@shijianjian thanks for the contribution. Would you like to volunteer with any of this missing items in the roadmap ? we are going to do a release this week and this two are missing yet: RandomResizedCrop
RandomCrop

shijianjian commented:
@edgarriba  Sure thing. I can help with that. I probably will do it tomorrow.
